### PR TITLE
8239345: need an enum for standard C types and a way to get a memory layout for those C types

### PR DESCRIPTION
**Pull request by sundararajana**
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8239345](https://bugs.openjdk.java.net/browse/JDK-8239345): need an enum for standard C types and a way to get a memory layout for those C types

### Download
`$ git fetch https://git.openjdk.java.net/panama-foreign pull/23/head:pull/23`
`$ git checkout pull/23`

### CODE REVIEW

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayouts.java [hunk 1]
```diff
@@ -144,139 +144,6 @@ private MemoryLayouts() {
      */
     public static final ValueLayout JAVA_DOUBLE = MemoryLayout.ofValueBits(64, ByteOrder.nativeOrder());
 
-    /**
-     * The {@code _Bool} native type.
-     */
-    public static final ValueLayout C_BOOL;
```
Comment (JornVernee) [inline src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayouts.java]: Can't we keep all these constants as well?

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayouts.java [hunk 2]
```diff
@@ -144,139 +144,6 @@ private MemoryLayouts() {
...
-    public static final ValueLayout C_DOUBLE;
-
-    /**
-     * The {@code T*} native type.
-     */
-    public static final ValueLayout C_POINTER;
-
-    static {
-        SystemABI abi = SystemABI.getInstance();
-        switch (abi.name()) {
```
Comment (JornVernee) [inline src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayouts.java]: We could use `SystemABI::layoutFor` here to do the initialization instead though.

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/aarch64/AArch64ABI.java [hunk 3]
```diff
@@ -72,5 +75,41 @@ public String name() {
         return SystemABI.ABI_AARCH64;
     }
 
-
+    @Override
+    public Optional<MemoryLayout> layoutFor(NativeType type) {
+        switch (type) {
```
Comment (JornVernee) [inline src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/aarch64/AArch64ABI.java]: Could use a switch expression here. Same for the other impls of SystemABI::layoutFor

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayouts.java (hunk 1 above)
Comment (sundararajana) [reply src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayouts.java]: I think the idea is to remove all C constants from MemoryLayouts.java (only keep Java layouts etc)

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayouts.java (hunk 2 above)
Comment (sundararajana) [reply src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayouts.java]: Please see above. Per discussion, the idea is to keep C-independent layout constants in MemoryLayouts.java.

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/SystemABI.java [hunk 4]
```diff
@@ -110,4 +110,96 @@ static SystemABI getInstance() {
         }
         throw new UnsupportedOperationException("Unsupported os or arch: " + os + ", " + arch);
     }
+
+    public enum NativeType {
```
[the remark this answers is not in the document]
Comment (sundararajana) [reply src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/SystemABI.java]: Agree. SystemABI.NativeType can be changed to SystemABI.Type

**Comment by sundararajana** [comment]
Will send fresh PR with changes suggested.

**Pull request closed without merging**